tls/s2n_resume.c

```diff
@@ -767,190 +767,157 @@ struct s2n_ticket_key *s2n_find_ticket_key(struct s2n_config *config, const uint
             S2N_ERR_KEY_CHECK);
-    POSIX_GUARD(s2n_stuffer_init(&aad, &aad_blob));
-    POSIX_GUARD(s2n_stuffer_write_bytes(&aad, key->implicit_aad, S2N_TICKET_AAD_IMPLICIT_LEN));
-    POSIX_GUARD(s2n_stuffer_write_bytes(&aad, key->key_name, S2N_TICKET_KEY_NAME_LEN));
 
-    uint32_t plaintext_header_size = s2n_stuffer_data_available(to);
-    POSIX_GUARD_RESULT(s2n_serialize_resumption_state(conn, to));
-    POSIX_GUARD(s2n_stuffer_skip_write(to, S2N_TLS_GCM_TAG_LEN));
+    /* Initialize Additional Authenticated Data */
+    uint8_t aad_data[S2N_TICKET_AAD_LEN] = { 0 };
+    struct s2n_blob aad_blob = { 0 };
+    RESULT_GUARD_POSIX(s2n_blob_init(&aad_blob, aad_data, sizeof(aad_data)));
+    struct s2n_stuffer aad = { 0 };
+    RESULT_GUARD_POSIX(s2n_stuffer_init(&aad, &aad_blob));
+    RESULT_GUARD_POSIX(s2n_stuffer_write_bytes(&aad, key->implicit_aad, sizeof(key->implicit_aad)));
+    RESULT_GUARD_POSIX(s2n_stuffer_write_bytes(&aad, key->key_name, sizeof(key->key_name)));
+
+    /* Write key name */
+    RESULT_GUARD_POSIX(s2n_stuffer_write_bytes(to, key->key_name, sizeof(key->key_name)));
+
+    /* Write IV */
+    uint8_t iv_data[S2N_TLS_GCM_IV_LEN] = { 0 };
+    struct s2n_blob iv = { 0 };
+    RESULT_GUARD_POSIX(s2n_blob_init(&iv, iv_data, sizeof(iv_data)));
+    RESULT_GUARD(s2n_get_public_random_data(&iv));
+    RESULT_GUARD_POSIX(s2n_stuffer_write(to, &iv));
 
+    /* Write serialized session state */
+    uint32_t plaintext_state_size = s2n_stuffer_data_available(to);
+    RESULT_GUARD(s2n_serialize_resumption_state(conn, to));
+    RESULT_GUARD_POSIX(s2n_stuffer_skip_write(to, S2N_TLS_GCM_TAG_LEN));
+
+    /* Initialize blob to be encrypted */
     struct s2n_blob state_blob = { 0 };
     struct s2n_stuffer copy_for_encryption = *to;
-    POSIX_GUARD(s2n_stuffer_skip_read(&copy_for_encryption, plaintext_header_size));
+    RESULT_GUARD_POSIX(s2n_stuffer_skip_read(&copy_for_encryption, plaintext_state_size));
     uint32_t state_blob_size = s2n_stuffer_data_available(&copy_for_encryption);
     uint8_t *state_blob_data = s2n_stuffer_raw_read(&copy_for_encryption, state_blob_size);
-    POSIX_ENSURE_REF(state_blob_data);
-    POSIX_GUARD(s2n_blob_init(&state_blob, state_blob_data, state_blob_size));
-
-    POSIX_GUARD(s2n_aes256_gcm.io.aead.encrypt(&aes_ticket_key, &iv, &aad_blob, &state_blob, &state_blob));
+    RESULT_ENSURE_REF(state_blob_data);
+    RESULT_GUARD_POSIX(s2n_blob_init(&state_blob, state_blob_data, state_blob_size));
 
-    POSIX_GUARD_RESULT(s2n_aes256_gcm.destroy_key(&aes_ticket_key));
+    RESULT_GUARD_POSIX(s2n_aes256_gcm.io.aead.encrypt(&aes_ticket_key, &iv, &aad_blob, &state_blob, &state_blob));
 
-    return S2N_SUCCESS;
+    return S2N_RESULT_OK;
 }
 
-int s2n_decrypt_session_ticket(struct s2n_connection *conn, struct s2n_stuffer *from)
+S2N_RESULT s2n_resume_decrypt_session_impl(struct s2n_connection *conn, struct s2n_stuffer *from,
+        uint64_t *key_intro_time)
```

Comment:
How about "s2n_resume_decrypt_session" instead? It's odd to have an "impl" without an corresponding non-"impl", and you should try to avoid "impl" if possible.

Also, probably mark this static.